### PR TITLE
Remove array allocation and local array ffi from the Show instance for records

### PR DESCRIPTION
**Pull request by ajnsit**
**Description of the change**

Removes array allocation from the implementation of show for records, and also consequently, removes the need for local ffi implementations of array functions just to show records.

---

**Checklist:**

- [X] Added the change to the changelog's "Unreleased" section with a reference to this PR (e.g. "- Made a change (#0000)")
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [X] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Data/Show.purs [hunk 1]
```diff
@@ -42,32 +48,45 @@ instance showArray :: Show a => Show (Array a) where
...
   ( Nub rs rs
   , RL.RowToList rs ls
   , ShowRecordFields ls rs
   ) =>
   Show (Record rs) where
   show record = case showRecordFields (Proxy :: Proxy ls) record of
-    [] -> "{}"
-    fields -> intercalate " " [ "{", intercalate ", " fields, "}" ]
+    "" -> "{}"
+    r -> "{ " <> r <> " }"
```
Comment (JordanMartinez) [inline src/Data/Show.purs]:
```purs
show record = "{" <> showRecordFields (Proxy :: Proxy ls) record) <> "}"
```

File: src/Data/Show.purs [hunk 2]
```diff
@@ -42,32 +48,45 @@ instance showArray :: Show a => Show (Array a) where
...
-  showRecordFields _ _ = []
-
+  showRecordFields _ _ = ""
+else
+instance showRecordFieldsConsNil ::
+  ( IsSymbol key
+  , Show focus
+  ) =>
+  ShowRecordFields (RL.Cons key focus RL.Nil) row where
+  showRecordFields _ record = key <> ": " <> show focus
```
Comment (JordanMartinez) [inline src/Data/Show.purs]:
```purs
" " <> key <> ": " <> show focus <> " "
```

File: src/Data/Show.purs [hunk 3]
```diff
@@ -42,32 +48,45 @@ instance showArray :: Show a => Show (Array a) where
...
+    focus = unsafeGet key record :: focus
+else
 instance showRecordFieldsCons ::
   ( IsSymbol key
   , ShowRecordFields rowlistTail row
   , Show focus
   ) =>
   ShowRecordFields (RL.Cons key focus rowlistTail) row where
-  showRecordFields _ record = cons (intercalate ": " [ key, show focus ]) tail
+  showRecordFields _ record = key <> ": " <> show focus <> ", " <> tail
```
Comment (JordanMartinez) [inline src/Data/Show.purs]:
```purs
" " <> key <> ": " <> show focus <> "," <> tail
```

**Review by JordanMartinez — COMMENTED** [review]
Correct me if I'm wrong, but can't this code...
```purs
show record = case showRecordFields (Proxy :: Proxy ls) record of
  "" -> "{}"
  r -> "{ " <> r <> " }"
```
be simplified to just
```purs
show record = "{" <> showRecordFields (Proxy :: Proxy ls) record) <> "}"
```
if the implementation is updated to the following?

File: src/Data/Show.purs (hunk 1 above)
Comment (ajnsit) [reply src/Data/Show.purs]: You are right, that does work and is a simpler way to accomplish the same thing. I have made the changes

**Review by garyb — APPROVED** [review]
Nice :+1:

**Review by JordanMartinez — APPROVED** [review]

File: src/Data/Show.purs [hunk 4]
```diff
@@ -20,6 +23,9 @@ import Type.Proxy (Proxy(..))
 class Show a where
   show :: a -> String
 
+instance showUnit :: Show Unit where
+  show _ = "unit"
+
```
Comment (JordanMartinez) [inline src/Data/Show.purs:28]: Could you clarify why this change was also made? And I guess the question is, should it be made?

Comment (garyb) [reply src/Data/Show.purs:28]: Same question for `Void` actually - maybe an earlier iteration of this change needed these in scope for its implementation?

Since all this is in the same library I guess it doesn't matter too much, but the general principle I used years ago when we split the `Prelude` into modules, was that only instances for `Prim` types or those that were unavoidably depended upon are introduced in the class module, otherwise they were introduced with the type.

Comment (ajnsit) [reply src/Data/Show.purs:28]: This was needed to avoid cycles in the import tree. I needed the semigroup instance for strings inside the show module, but `Data.Semigroup` depends on `Data.Void` and `Data.Unit`, which in turn depended on `Data.Show` for defining their show instance.

The rationale for why I resolved the cyclic imports this way is - by removing typeclass dependencies from "core" data structures like `Unit` and `Void`, we can trivially ensure they never cause cyclical imports and any typeclasses can always import the data structures. This way we only need to organise and resolve potential cyclical imports between typeclass modules like `Data.Show` and `Data.Semigroup`.

Comment (JordanMartinez) [reply src/Data/Show.purs:28]: Gotcha. Thanks for clarifying!
